### PR TITLE
[SPARK-8473][SPARK-9889][ML] User guide and example code for DCT

### PR DESCRIPTION
**Pull request by feynmanliang**
@mengxr @jkbradley

### CODE REVIEW

File: docs/ml-features.md [hunk 1]
```diff
@@ -649,6 +649,70 @@ for expanded in polyDF.select("polyFeatures").take(3):
 </div>
 </div>
 
+## Discrete Cosine Transform (DCT)
+
+The [Discrete Cosine Transform](https://en.wikipedia.org/wiki/Discrete_cosine_transform) transforms a length $N$ real-valued sequence in the time domain into another length $N$ real-valued sequence in the frequency domain. A [DCT](api/scala/index.html#org.apache.spark.ml.feature.DCT) class provides this functionality, implementing the [DCT-II](https://en.wikipedia.org/wiki/Discrete_cosine_transform#DCT-II) and scaling the result by $1/\sqrt{2}$ such that the representing matrix for the transform is unitary. No shift is applied to the transformed sequence (e.g. the $0$th element of the transformed sequence is the $0$th DCT coefficient and _not_ the $N/2$th).
```
Comment (mengxr) [inline docs/ml-features.md]: minor: Please maintain the line width at 100 and end sentences with new lines. This helps reduce the diff size of future edits.

File: docs/ml-features.md [hunk 2]
```diff
@@ -649,6 +649,70 @@ for expanded in polyDF.select("polyFeatures").take(3):
...
+{% highlight scala %}
+import org.apache.spark.ml.feature.DCT
+import org.apache.spark.mllib.linalg.Vectors
+
+val data = Seq(
+  Vectors.dense(0.0, 1.0, -2.0, 3.0),
+  Vectors.dense(-1.0, 2.0, 4.0, -7.0),
+  Vectors.dense(14.0, -2.0, -5.0, 1.0))
+val df = sqlContext.createDataFrame(data.map(Tuple1.apply)).toDF("features")
+val DCTransform = new DCT()
```
Comment (mengxr) [inline docs/ml-features.md]: `DCTransform` -> `dct`? (more consistent with naming convention)

File: docs/ml-features.md [hunk 3]
```diff
@@ -649,6 +649,70 @@ for expanded in polyDF.select("polyFeatures").take(3):
...
+  .setInputCol("features")
+  .setOutputCol("featuresDCT")
+  .setInverse(false)
+val DCTdf = DCTransform.transform(df)
```
Comment (mengxr) [inline docs/ml-features.md]: `DCTdf` -> `dctDf`

File: docs/ml-features.md [hunk 4]
```diff
@@ -649,6 +649,70 @@ for expanded in polyDF.select("polyFeatures").take(3):
...
+import org.apache.spark.mllib.linalg.Vectors;
+import org.apache.spark.sql.DataFrame;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.RowFactory;
+import org.apache.spark.sql.SQLContext;
+import org.apache.spark.sql.types.Metadata;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+
+JavaRDD<Row> data = jsc.parallelize(Lists.newArrayList(
```
Comment (mengxr) [inline docs/ml-features.md]: minor: `Lists.newArrayList` -> `Arrays.asList` (to remove Guava dependency)

File: docs/ml-features.md (hunk 1 above)
Comment (feynmanliang) [inline docs/ml-features.md]: OK

File: docs/ml-features.md (hunk 2 above)
Comment (feynmanliang) [inline docs/ml-features.md]: OK

File: docs/ml-features.md (hunk 3 above)
Comment (feynmanliang) [inline docs/ml-features.md]: OK

File: docs/ml-features.md (hunk 4 above)
Comment (feynmanliang) [inline docs/ml-features.md]: OK. I've updated this PR and created a starter JIRA at SPARK-10070 to do this across all user guide

File: docs/ml-features.md [hunk 5]
```diff
@@ -649,6 +649,80 @@ for expanded in polyDF.select("polyFeatures").take(3):
...
+  Vectors.dense(0.0, 1.0, -2.0, 3.0),
+  Vectors.dense(-1.0, 2.0, 4.0, -7.0),
+  Vectors.dense(14.0, -2.0, -5.0, 1.0))
+val df = sqlContext.createDataFrame(data.map(Tuple1.apply)).toDF("features")
+val dct = new DCT()
+  .setInputCol("features")
+  .setOutputCol("featuresDCT")
+  .setInverse(false)
+val dctDf = dct.transform(df)
+dctDf.select("featuresDCT").take(3).foreach(println)
```
Comment (jkbradley) [inline docs/ml-features.md]: ```dctDf.select("featuresDCT").show(3)``` gives a nicer format

File: docs/ml-features.md [hunk 6]
```diff
@@ -649,6 +649,80 @@ for expanded in polyDF.select("polyFeatures").take(3):
...
+import org.apache.spark.sql.DataFrame;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.RowFactory;
+import org.apache.spark.sql.SQLContext;
+import org.apache.spark.sql.types.Metadata;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+
+JavaRDD<Row> data = jsc.parallelize(Arrays.asList(
+      RowFactory.create(Vectors.dense(0.0, 1.0, -2.0, 3.0)),
```
Comment (jkbradley) [inline docs/ml-features.md]: 2 space indentation

**Comment by jkbradley** [comment]
Looks good except for those items.  I ran the Scala example.

File: docs/ml-features.md (hunk 6 above)
Comment (feynmanliang) [inline docs/ml-features.md]: OK

File: docs/ml-features.md [hunk 7]
```diff
@@ -649,6 +649,77 @@ for expanded in polyDF.select("polyFeatures").take(3):
...
+StructType schema = new StructType(new StructField[] {
+  new StructField("features", new VectorUDT(), false, Metadata.empty()),
+});
+DataFrame df = jsql.createDataFrame(data, schema);
+DCT dct = new DCT()
+  .setInputCol("features")
+  .setOutputCol("featuresDCT")
+  .setInverse(false);
+DataFrame dctDf = dct.transform(df);
+dctDf.select("featuresDCT").take(3).show(3);
```
Comment (jkbradley) [inline docs/ml-features.md]: Remove "take(3)"

Comment (feynmanliang) [inline docs/ml-features.md]: OK

**Comment by jkbradley** [comment]
LGTM merging with master and branch-1.5

**Pull request closed without merging**